projects/pgai/pgai/alembic/operations.py

```diff
@@ -0,0 +1,248 @@
+    params = [
+        f"'{config.model}'",
+        str(config.dimensions),
+    ]
+    if config.chat_user:
+        params.append(f"chat_user=>'{config.chat_user}'")
+    if config.api_key_name:
+        params.append(f"api_key_name=>'{config.api_key_name}'")
+    return f"ai.embedding_openai({', '.join(params)})"
+
+
+def _build_chunking_params(config: ChunkingConfig) -> str:
+    """Build chunking configuration parameters."""
+    params = [f"'{config.chunk_column}'"]
+    if config.chunk_size is not None:
+        params.append(f"chunk_size=>{config.chunk_size}")
+    if config.chunk_overlap is not None:
+        params.append(f"chunk_overlap=>{config.chunk_overlap}")
+    if config.separator is not None:
+        if isinstance(config.separator, list):
+            sep_str = "array[" + ",".join(f"'{s}'" for s in config.separator) + "]"
+        else:
+            sep_str = f"array['{config.separator}']"
+        params.append(f"separators=>{sep_str}")
+    if config.is_separator_regex:
+        params.append("is_separator_regex=>true")
+    return f"ai.chunking_recursive_character_text_splitter({', '.join(params)})"
+
+def _build_diskann_indexing_params(config: DiskANNIndexingConfig) -> str:
+    """Build DiskANN indexing configuration parameters."""
+    params = []
+    if config.min_rows is not None:
+        params.append(f"min_rows=>{config.min_rows}")
+    if config.storage_layout is not None:
+        params.append(f"storage_layout=>'{config.storage_layout}'")
+    if config.num_neighbors is not None:
+        params.append(f"num_neighbors=>{config.num_neighbors}")
+    if config.search_list_size is not None:
+        params.append(f"search_list_size=>{config.search_list_size}")
+    if config.max_alpha is not None:
+        params.append(f"max_alpha=>{config.max_alpha}")
+    if config.num_dimensions is not None:
+        params.append(f"num_dimensions=>{config.num_dimensions}")
+    if config.num_bits_per_dimension is not None:
+        params.append(f"num_bits_per_dimension=>{config.num_bits_per_dimension}")
+    if config.create_when_queue_empty is not None:
+        params.append(f"create_when_queue_empty=>{str(config.create_when_queue_empty).lower()}")
+    return f"ai.indexing_diskann({', '.join(params)})"
+
+
+def _build_hnsw_indexing_params(config: HNSWIndexingConfig) -> str:
+    """Build HNSW indexing configuration parameters."""
+    params = []
+    if config.min_rows is not None:
+        params.append(f"min_rows=>{config.min_rows}")
+    if config.opclass is not None:
+        params.append(f"opclass=>'{config.opclass}'")
+    if config.m is not None:
+        params.append(f"m=>{config.m}")
+    if config.ef_construction is not None:
+        params.append(f"ef_construction=>{config.ef_construction}")
+    if config.create_when_queue_empty is not None:
+        params.append(f"create_when_queue_empty=>{str(config.create_when_queue_empty).lower()}")
+    return f"ai.indexing_hnsw({', '.join(params)})"
```

Comment:
One thing I am considering is to make the dataclass config objects smarter:
Currently there is multiple places where logic for each config is placed:
- rendering to SQL (here)
- rendering to python (in autogenerate.py)
- loading from the db state (in compare_vectorizers)
- Storing configuration in the metadata info dict on the vectorizer field

I think if I make these config objects "smarter" and add a few functions:
- `from_db_state`
- `to_sql()`
- `to_python` or just `__repr__`?

For each one. It should make keeping this in sync with our actual underlying db code a lot easier. Since you'll only have to touch one spot in the code and not understand how everything works?

Honestly I am a bit disappointed by alembic, i would have thought that at least the `to_python` part is covered by the library generically already.

Reply:
Note: this actually kind of overlaps with the pydantic models we have in the vectorizer part of the library. Maybe there is also a way to re-use that config. But e.g. `IndexingConfig` does not exist there and needs to be loaded separately.

Reply:
I strongly encourage us to have one set of dataclass objects across all of our libraries. That way we have code reuse and only one place to add things. (and I expect we'll constantly be adding things). My mental model would be one set of dataclass objects (probably pydantic based) that's used (imported) by both vectorizer and these kind of integrations. So that could include things like `IndexingConfig` that's used by only some parts of the code and not others.

Reply:
It might actually be nice to have access to them in the extension too. I could use the pydantic models to validate the jsonb representations. But I don't know how cumbersome it would be to do this.